pkg/auto/oci.go

```diff
@@ -0,0 +1,192 @@
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Pushed assets: %v", pushedAssets)
+
+	return nil
+}
+
+func setupOCI(ociDNS, ociUser, ociPass string, debug bool) (*oci, error) {
+	var err error
+	o := &oci{
+		DNS:      ociDNS,
+		user:     ociUser,
+		password: ociPass,
+	}
+
+	o.helmClient, err = setupHelm(o.DNS, o.user, o.password, debug)
+	if err != nil {
+		return nil, err
+	}
+
+	o.loadAsset = loadAsset
+	o.checkAsset = checkAsset
+	o.push = push
+
+	return o, nil
+}
+
+func setupHelm(ociDNS, ociUser, ociPass string, debug bool) (*registry.Client, error) {
+	settings := cli.New()
+	actionConfig := new(action.Configuration)
+	if err := actionConfig.Init(settings.RESTClientGetter(), settings.Namespace(), os.Getenv("HELM_DRIVER"), func(format string, v ...interface{}) {
+		fmt.Sprintf(format, v...)
+	}); err != nil {
+		return nil, err
+	}
+
+	var regClient *registry.Client
+
+	if debug {
+		fmt.Println("debug mode you need to provide a self-signed certificate")
+		caFile := "/etc/docker/certs.d/" + ociDNS + "/ca.crt"
+
+		regClient, err := registry.NewRegistryClientWithTLS(os.Stdout, "", "", caFile, false, "", true)
+		if err != nil {
+			return nil, err
+		}
+
+		if err := regClient.Login(
+			ociDNS,
+			registry.LoginOptInsecure(false),
+			registry.LoginOptTLSClientConfig("", "", caFile),
+			registry.LoginOptBasicAuth(ociUser, ociPass),
+		); err != nil {
+			return nil, err
+		}
+
+		return regClient, nil
+	}
+
+	regClient, err := registry.NewClient(registry.ClientOptDebug(false))
+	if err != nil {
+		return nil, err
+	}
+
+	if err := regClient.Login(ociDNS,
+		registry.LoginOptInsecure(false),
+		registry.LoginOptBasicAuth(ociUser, ociPass)); err != nil {
+		return nil, err
+	}
+
+	return regClient, nil
+}
+
+func (o *oci) update(release *options.ReleaseOptions) ([]string, error) {
+	var pushedAssets []string
+
+	for chart, versions := range *release {
+		for _, version := range versions {
+			asset := chart + "-" + version + ".tgz"
+			assetData, err := o.loadAsset(chart, asset)
+			if err != nil {
+				fmt.Printf("Failed to load asset: %s; error: %s \n", asset, err.Error())
+				return pushedAssets, err
+			}
+
+			// Check if the asset version already exists in the OCI registry
+			// Never overwrite a previously released chart!
+			exists, err := o.checkAsset(o.helmClient, o.DNS, chart, version)
+			if err != nil {
+				fmt.Printf("Failed to check registry for %s; error: %s \n", asset, err.Error())
+				return pushedAssets, err
+			}
+			if exists {
+				fmt.Printf("Asset %s already exists in the OCI registry \n", asset)
+				return pushedAssets, fmt.Errorf("asset %s already exists in the OCI registry", asset)
+			}
+
+			fmt.Printf("Pushing asset to OCI Registry: %s \n", asset)
+			if err := o.push(o.helmClient, assetData, buildPushURL(o.DNS, chart, version)); err != nil {
+				err = fmt.Errorf("failed to push %s; error: %w ", asset, err)
+				fmt.Printf(err.Error())
+				return pushedAssets, err
```

Comment:
It would be better if we could make an atomic transaction and push all assets or push none, but if this isn't possible, you should do the same as you're doing in the last err check block, composing the errors instead of just logging them